File: solutions/julia/protein-translation/protein-translation.jl
function proteins(strand)
    codon_map = Dict(
        "AUG" => "Methionine",
        "UUU" => "Phenylalanine",
        "UUC" => "Phenylalanine",
        "UUA" => "Leucine",
        "UUG" => "Leucine",
        "UCU" => "Serine",
        "UCC" => "Serine",
        "UCA" => "Serine",
        "UCG" => "Serine",
        "UAU" => "Tyrosine",
        "UAC" => "Tyrosine",
        "UGU" => "Cysteine",
        "UGC" => "Cysteine",
        "UGG" => "Tryptophan",
        "UAA" => "STOP",
        "UAG" => "STOP",
        "UGA" => "STOP"
    )

    amino_acids = []

    for i in 1:3:length(strand)
        codon = strand[i:min(i+2, length(strand))]

        if length(codon) != 3
            throw(DomainError(""))
        end

        if haskey(codon_map, codon)
            if codon_map[codon] == "STOP"
                return amino_acids
            else
                push!(amino_acids, codon_map[codon])
            end
        else
            throw(DomainError(""))
        end
    end

    return amino_acids

end

File: solutions/cpp/protein-translation/protein_translation.cpp
#include "protein_translation.h"

#include <map>
#include <stdexcept>

namespace protein_translation {

    std::map<std::string, std::string> codon_map = {
        {"AUG", "Methionine"},
        {"UUU", "Phenylalanine"},
        {"UUC", "Phenylalanine"},
        {"UUA", "Leucine"},
        {"UUG", "Leucine"},
        {"UCU", "Serine"},
        {"UCC", "Serine"},
        {"UCA", "Serine"},
        {"UCG", "Serine"},
        {"UAU", "Tyrosine"},
        {"UAC", "Tyrosine"},
        {"UGU", "Cysteine"},
        {"UGC", "Cysteine"},
        {"UGG", "Tryptophan"},
        {"UAA", "STOP"},
        {"UAG", "STOP"},
        {"UGA", "STOP"}
    };

    auto proteins(const std::string &strand) -> std::vector<std::string>
    {
        std::vector<std::string> amino_acids;

        for (std::size_t i = 0; i < strand.size() - 2; i += 3)
        {
            auto codon = strand.substr(i, 3);

            if (codon.size() != 3)
            {
                return amino_acids;
            }

            if (codon_map.find(codon) != codon_map.end())
                if (codon_map[codon] == "STOP")
                    return amino_acids;
                else
                    amino_acids.push_back(codon_map[codon]);
            else
                throw(std::domain_error(""));
        }

        return amino_acids;
    }

}  // namespace protein_translation
